### DISCUSSION TITLE
Issue: Script Helper Tool with the same experiment name overwrites previous files without warning

### DISCUSSION THREAD

**Discussion by A-Ivan** [BioMotionLab/TUX#53] [Ideas]
Issue #44 

"If you use the **Script Helper Tool** to setup the experiment automatically more than once, with the same name, it will overwrite any previously created and edited files.

I suggest adding a function to check if these files already exist and to provide a popup of some kind asking the user to confirm if they want to overwrite the existing files.

Confirming would recreate/overwrite the files experiment files.

Canceling would keep the user at the Script Helper window without overwriting any files. The user could then change the experiment name and then create/setup the desired files."

My idea is to solve this is to check if there are already existing files, when one of the buttons in the Script Helper Tool window are pressed, and if so, show a popup window which asks if the user would like to overwrite existing files or cancel the file creation.

To get this to work, I need to understand how to access the Script Helper Tool script and its button calls - I will need to add checks to the button press functions for scripts with the "experiment_name_script_type". For instance, if the user's experiment is called "Experiment" and the Runner script button is pressed, search for a folder with the name experiment name and then search for the ExperimentRunner script in that folder.

**Comment by AdamBebko** [comment]
This is where all the helper tool stuff is. That one should be a fairly easy fix.

![Screen Shot 2022-01-20 at 7 04 39 PM](https://user-images.githubusercontent.com/43276915/150441681-dd92f023-1dbb-4fca-b0da-774ed8a2a313.png)

**Comment by A-Ivan** [comment]
Hi @AdamBebko

I was able to get this working but I wasn't able to push to the dev branch or create a new branch and create a pull request, it says that permission is denied.

**Summary of the changes:**

1. Created a **ScriptHelperErrorEditorWindow** class, which inherits the EditorWindow. This will appears when there is an already existing file which the system is trying to overwrite.
2. **ScriptHelperEditorWindow** has been changed. I have added a boolean called "allowFileOverwrite" (which is drawn as a toggle in the Script Helper Window). If this is set to true (checked) files can be overwritten when the button to create the respective file is pressed, if false (not checked) files will not be overwritten and the "ScriptHelperErrorEditorWindow" Editor window will appear informing the user that the file already exists and that they need to set the boolean as true in the Script Helper Window if they want to overwrite it.

The current solution is a quick one, but may not be the most elegant.

What I wanted to do originally, check if the file already exists and if it does, then ask the user if they want to overwrite it via a popup menu. The problem with this is that I was not able to send the user's button choice (to overwrite or not the file) back to the respective ScriptHelperEditorWindow's ScriptTemplateComponent. I thought of sending the component's name to the popup window, but then the code was getting complicated and ugly.

This is what the Script Helper Window currently looks like:

![image](https://user-images.githubusercontent.com/17125077/151617475-35d843c3-0291-44ce-9401-fe93a87c476b.png)

This is what the "ScriptHelperErrorEditorWindow" looks like:

![image](https://user-images.githubusercontent.com/17125077/151618854-80bbbd7f-92d0-47d4-b53e-9c7874aca108.png)

If you have any questions about anything let me know.

**Comment by A-Ivan** [comment]
Note, when the "Create Experiment Design File" button is pressed, it creates both the Design and the Settings Files, not just the Design File.

Do you want me to separate the Design file creation from creating both of those files and only creating the Design file? I would also then add the "new SettingsExperimentAsset().CreateAsset(experimentName, allowFileOverwrite);" to the "Automatically set everything up for me!" button function list.

Not sure if this would change too much of how things work internally for your code.

**Reply by AdamBebko** [reply]
I'll look into the permissions problem and take a look at these modifications this week. Thanks.

**Reply by A-Ivan** [reply]
Great, you're welcome.

**Reply by AdamBebko** [reply]
These modifications look great. One option to make the popup work as you wanted would be to create another class that actually displays the popup, and retrieves the result. When creating this popup class you could pass it a callback `Action` to call on closing of the popup. When calling the action, through Invoke() you could pass it the result of the popup function. eg: `Invoke(shouldOverwrite);`

**Reply by AdamBebko** [reply]
Again, try doing the pull request from a fork.
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork
